### DISCUSSION TITLE
Proper way to apply patch to BPMP Device Tree Blobs

### DISCUSSION THREAD

**Discussion by dwalkes** [OE4T/meta-tegra#1094] [Q&A] [answered]
Hi,
I'm looking to patch the BPMPFW DTB as described at [this link](https://docs.nvidia.com/jetson/archives/l4t-archived/l4t-3261/index.html#page/Tegra%2520Linux%2520Driver%2520Package%2520Development%2520Guide/clocks.html%23wwpID0E04B0HA) to use the PLLAON as a clock source for better CAN bitrate accuracy.

I've hacked a solution into a `tegra-bootfiles_%.bbappend` in my layer which looks like this and converts each dtb file to a `-pllaon-enabled` dtb by applying patches:
```
PATCH_PATH := "${THISDIR}/tegra-bootfiles/"

do_compile_append_tegra194() {
    for dtbfile in $(ls ${S}/bootloader/${NVIDIA_BOARD}/tegra194-*-bpmp-*.dtb); do
        if echo "$dtbfile" | grep -q "pllaon-enabled"; then
            continue
        fi
        dtbdir=$(dirname $dtbfile)
        dtsfile=$(basename $dtbfile .dtb).dts
        pllaondts=$(basename $dtbfile .dtb)-pllaon-enabled.dts
        pllaondtb=$(basename $dtbfile .dtb)-pllaon-enabled.dtb
        dtc -I dtb -O dts ${dtbfile} -o ${pllaondts}
        patchfile=${PATCH_PATH}/${pllaondts}.patch
        if echo "$dtbfile" | grep -q "tegra194-a02-bpmp"; then
            patchfile=${PATCH_PATH}/tegra194-a02-bpmp-p2888-a01-pllaon-enabled.dts.patch
        fi
        if echo "$dtbfile" | grep -q "tegra194-a01-bpmp"; then
            patchfile=${PATCH_PATH}/tegra194-a01-bpmp-p2888-a01-pllaon-enabled.dts.patch
        fi
        if [ ! -e ${patchfile} ]; then
            echo "Missing patch file ${pllaondts}.patch.  Please add a patch to enable pllon for ${dtsfile}"
            exit 1
        fi

        if echo "$dtsfile" | grep -q "tegra194-a02-bpmp-p2888-0008-b01-jaxi"; then
            echo "pllaon already enabled on ${dtsfile}, no patch step required"
            ln -sf ${dtbfile} ${dtbdir}/${pllaondtb}
        else
            patch -i ${patchfile} ${pllaondts}
            dtc -I dts -O dtb ${pllaondts} -o ${dtbdir}/${pllaondtb}
        fi
    done
}
```

Where the patch files look like:
```
--- tegra194-a01-bpmp-p2888-a01-pllaon-enabled.dts.orig	2022-10-05 12:16:15.398539490 -0500
+++ tegra194-a01-bpmp-p2888-a01-pllaon-enabled.dts	2022-10-05 12:38:06.490878683 -0500
@@ -22,7 +22,7 @@
 
 		clock@can1 {
 			allow_fractional_divider = <0x01>;
-			allowed-parents = <0x121 0x5b>;
+			allowed-parents = <0x121 0x5b 0x5e>;
 			clk-id = <0x09>;
 		};
 
@@ -36,7 +36,7 @@
 
 		clock@can2 {
 			allow_fractional_divider = <0x01>;
-			allowed-parents = <0x121 0x5b>;
+			allowed-parents = <0x121 0x5b 0x5e>;
 			clk-id = <0x0b>;
 		};

```

I then use this append on tegra-flashvars to redirect to use the `-pllaon-enabled` version of the .dst.
```
do_install_append_tegra194() {
    echo "Append pllaon-enabled version fo BPFDTB_FILE to flashvars to use the pllaon enabled bpmp config"
    sed -i 's/^\(BPFDTB_FILE=\)\(.*\)\(.dtb\)/\1\2-pllaon-enabled\3/g' ${D}${datadir}/tegraflash/flashvars 
}
```

However, I'm guessing the `PATCH_PATH := "${THISDIR}/tegra-bootfiles/"` hack is probably not the best way to do this.

Since [tegra-bootfiles requires tegra-shared-binaries.inc](https://github.com/OE4T/meta-tegra/blob/e4aba7671cfcffea2ca5212579fd0986c003d426/recipes-bsp/tegra-binaries/tegra-bootfiles_35.1.0.bb#L2) and [tegra-shared-binaries disables do_fetch](https://github.com/OE4T/meta-tegra/blob/master/recipes-bsp/tegra-binaries/tegra-shared-binaries.inc#L1-L4) I can't use the typical SRC_URI method of referencing/using these patches.  I thought about using my own recipe but if possible it would be nice to use an append on tegra-bootfiles so I could continue to reuse the [do_install logic there](https://github.com/OE4T/meta-tegra/blob/e4aba7671cfcffea2ca5212579fd0986c003d426/recipes-bsp/tegra-binaries/tegra-bootfiles_35.1.0.bb#L99-L101).

I'm wondering if someone has already solved this problem in a better way or has some suggestions about a better way to handle this.

Thanks!

**Comment by kekiefer** [comment]
It's been a while since I did this, but I think I ran into something similar, patching the `tegra-mb1-soft-fuses-l4t.cfg` file. I ended up patching it in the recipe for `tegra-binaries`, where it was picked up later by other recipes. I'm not 100% sure if this will work for you, but it's worth a look.

**Comment by madisongh** [comment, marked as answer]
I'm not sure the usual SRC_URI/do_patch mechanism would work for this anyway, since the BPMP device trees are provided in binary form only, so you'd still have to decompile them, apply the patches, then recompile them.  You could add the patches to the SRC_URI in the `tegra-binaries` recipes as @kekiefer suggests - but make sure to tag them with `apply=no` so they don't get processed by do_patch.

Doing this in a separate recipe wouldn't be too bad, either -- have `DEPENDS = "tegra-bootfiles"` in the recipe, then extract the original dtb files from the recipe sysroot, process them, then install the new ones back in the recipe sysroot again.  The drawback with that approach is making sure you have dependencies in the right places for the added recipe.

I suppose we could split the `tegra-bootfiles` recipe in two, like we've done for some of the others... create a `tegra-bootfiles-base` recipe that does what the current one does, then have `tegra-bootfiles` (which by default wouldn't do anything on its own) that depends on the `-base` recipe.  That would make it easier to add customizations like this, since the outer recipe wouldn't have the restrictions on it that come from the `tegra-shared-binaries` stuff.

**Reply by dwalkes** [reply]
> I'm not sure the usual SRC_URI/do_patch mechanism would work for this anyway, since the BPMP device trees are provided in binary form only, so you'd still have to decompile them, apply the patches, then recompile them

Yea that makes sense.  I was mostly just looking for a solution to the fetch part, not the patch.  Thanks for the tip on `apply=no`.

> I suppose we could split the tegra-bootfiles recipe in two, like we've done for some of the others... create a tegra-bootfiles-base recipe that does what the current one does, then have tegra-bootfiles (which by default wouldn't do anything on its own) that depends on the -base recipe. That would make it easier to add customizations like this, since the outer recipe wouldn't have the restrictions on it that come from the tegra-shared-binaries stuff.

I'll look into this if anyone else is interested in using it, otherwise will proceed with the `tegra-binaries` + `apply=no` route.
